### PR TITLE
In validate function alter option strict to allowed, with option 'strict' and 'warn'

### PR DESCRIPTION
**Pull request by benTrust**
Alter option strict to allowed, with parameters 'strict' and 'warn'

### CODE REVIEW

**Comment by madarche** [comment]
It's a bit sad that the code coverage decreased. @benTrust could you do a `npm run test:coverage` and check in the HTML report what parts of the new code that you have contributed is in red and thus not covered by tests? To increase the code coverage, you should then modify existing tests or create new tests to be sure that all the possibilities are covered. It seems that you have not covered the `{allowed: 'warn'}` case in the tests.

**Comment by madarche** [comment]
@benTrust good for the code coverage increase 👍

File: README.md [hunk 1]
```diff
@@ -65,7 +65,7 @@ var env = conf.get('env');
 conf.loadFile('./config/' + env + '.json');
 
 // Perform validation
-conf.validate({strict: true});
+conf.validate(allowed: 'strict');
```
Comment (madarche) [inline README.md]: Error here: it should be conf.validate({allowed: 'strict'});

File: README.md [hunk 2]
```diff
@@ -352,14 +352,16 @@ conf.loadFile(process.env.CONFIG_FILES.split(','));
...
 
 Validates `config` against the schema used to initialize it. All errors are
-collected and thrown at once.
+collected and thrown or displayed at once.
 
-Options: At the moment `strict` is the only available option.
-
-If the `strict` option is passed (that is `{strict: true}` is passed), any
+#### Options:
+1. `strict`: If the `strict` option is passed (that is `{allowed: 'strict'}` is passed), any
```
Comment (madarche) [inline README.md]: `strict` is not an option. `allowed` is. `strict` is just a possible value of the `allowed` option

File: README.md [hunk 3]
```diff
@@ -352,14 +352,16 @@ conf.loadFile(process.env.CONFIG_FILES.split(','));
...
 properties specified in config files that are not declared in the schema will
 result in errors. This is to ensure that the schema and the config files are in
-sync. By default the strict mode is set to false.
+sync.
+2. `warn`: If the `warn` option is passed (that is `{allowed: 'warn'}` is passed), any
```
Comment (madarche) [inline README.md]: `warn` is not an option. `allowed` is. `warn` is just a possible value of the `allowed` option

File: lib/convict.js [hunk 4]
```diff
@@ -501,27 +508,53 @@ let convict = function convict(def) {
      */
     validate: function(options) {
       options = options || {};
-      options.strict = options.strict || false;
-      let errors = validate(this._instance, this._schema, [],options.strict);
 
-      if (errors.length) {
-        let errBuf = '';
+      if ('strict' in options) {
```
Comment (madarche) [inline lib/convict.js]: Put `strict` as a `ALLOWED_OPTION_STRICT` `const` somewhere the top of the file

File: lib/convict.js [hunk 5]
```diff
@@ -501,27 +508,53 @@ let convict = function convict(def) {
...
+              err_buf +=  ': value was ' + JSON.stringify(e.value);
+            }
           }
+          return err_buf;
+        };
+
+        let types_err_buf = fillErrorBuffer(errors.values_not_good_type);
+        let params_err_buf = fillErrorBuffer(errors.params_not_declared);
+
+        if (options.allowed === 'warn' && params_err_buf.length) {
```
Comment (madarche) [inline lib/convict.js]: Put `warn` as a `ALLOWED_OPTION_WARN` `const` somewhere the top of the file

**Comment by benTrust** [comment]
I modify the message of my branch commit ( 'In validate function alter option strict to allowed, with OPTIONS..' to 'In validate function alter option strict to allowed, with VALUES...') and follow the @madarche  advices

File: lib/convict.js [hunk 6]
```diff
@@ -501,27 +511,53 @@ let convict = function convict(def) {
      */
     validate: function(options) {
       options = options || {};
-      options.strict = options.strict || false;
-      let errors = validate(this._instance, this._schema, [],options.strict);
 
-      if (errors.length) {
-        let errBuf = '';
+      if (ALLOWED_OPTION_STRICT in options) {
```
Comment (madarche) [inline lib/convict.js]: This is an error here: `ALLOWED_OPTION_STRICT` cannot be in `options`. `ALLOWED_OPTION_STRICT` can only be in `options.allowed`.

You have done one search-and-replace too many! Here it should be: `if ('strict' in options)`.

**Comment by madarche** [comment]
Apart from this last detail, the whole looks good. I'm looking forward to merge this!

**Comment by madarche** [comment]
This seems all perfect to me 🥇 
Thank you a lot for all the good work

**Pull request merged**